Bottom: c8e2f573c4c0a18e335eec677c6d0f3ccd900d7f
Top:    7e93c56ca35ea9a24ff811af9573108b5687e393
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-25 13:15:45 +0000

Fixup Z-heights of exported copper


---

diff --git a/src/hid/common/object3d.c b/src/hid/common/object3d.c
index c1ed686..3b48da6 100644
--- a/src/hid/common/object3d.c
+++ b/src/hid/common/object3d.c
@@ -254,8 +254,6 @@ get_contour_edge_n_is_round (PLINE *contour, int n)
     }
 #endif
 
-  return false;
-
   while (n > 0) {
     edge = edge->next; /* The VNODE structure is circularly linked, so wrapping is OK */
     n--;
@@ -696,14 +694,14 @@ object3d_from_board_outline (void)
 
 #ifdef REVERSED_PCB_CONTOURS
   objects = object3d_from_contours (board_outline,
-                                    -HACK_BOARD_THICKNESS + HACK_MASK_THICKNESS, /* Bottom */
-                                    0                     -HACK_MASK_THICKNESS,  /* Top */
+                                    -HACK_BOARD_THICKNESS, /* Bottom */
+                                    0                    ,  /* Top */
                                     board_appearance,
                                     top_bot_appearance);
 #else
   objects = object3d_from_contours (board_outline,
-                                     HACK_BOARD_THICKNESS / 2 - HACK_MASK_THICKNESS, /* Bottom */
-                                    -HACK_BOARD_THICKNESS / 2 + HACK_MASK_THICKNESS, /* Top */
+                                     HACK_BOARD_THICKNESS / 2, /* Bottom */
+                                    -HACK_BOARD_THICKNESS / 2, /* Top */
                                     board_appearance,
                                     top_bot_appearance);
 #endif
@@ -963,14 +961,14 @@ object3d_from_soldermask_within_area (POLYAREA *area, int side)
 
 #ifdef REVERSED_PCB_CONTOURS
   objects = object3d_from_contours (info.poly,
-                                    (side == TOP_SIDE) ? -HACK_MASK_THICKNESS : -HACK_BOARD_THICKNESS,                       /* Bottom */
-                                    (side == TOP_SIDE) ? 0                    : -HACK_BOARD_THICKNESS + HACK_MASK_THICKNESS, /* Top */
+                                    (side == TOP_SIDE) ? 0                   - HACK_COPPER_THICKNESS : -HACK_BOARD_THICKNESS - HACK_COPPER_THICKNESS - HACK_MASK_THICKNESS, /* Bottom */
+                                    (side == TOP_SIDE) ? HACK_MASK_THICKNESS - HACK_COPPER_THICKNESS : -HACK_BOARD_THICKNESS - HACK_COPPER_THICKNESS,                       /* Top */
                                     mask_appearance,
                                     NULL);
 #else
   objects = object3d_from_contours (info.poly,
-                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 + HACK_MASK_THICKNESS : HACK_BOARD_THICKNESS / 2,                       /* Bottom */
-                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2                       : HACK_BOARD_THICKNESS / 2 - HACK_MASK_THICKNESS, /* Top */
+                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 - HACK_COPPER_THICKNESS                       : HACK_BOARD_THICKNESS / 2 + HACK_COPPER_THICKNESS + HACK_MASK_THICKNESS, /* Bottom */
+                                    (side == TOP_SIDE) ? -HACK_BOARD_THICKNESS / 2 - HACK_COPPER_THICKNESS - HACK_MASK_THICKNESS : HACK_BOARD_THICKNESS / 2 + HACK_COPPER_THICKNESS, /* Top */
                                     mask_appearance,
                                     NULL);
 #endif
@@ -1335,7 +1333,7 @@ compute_depth (int group)
   int min_copper_group;
   int max_copper_group;
   int num_copper_groups;
-  int middle_copper_group;
+//  int middle_copper_group;
   int depth;
 
   top_group = GetLayerGroupNumberBySide (TOP_SIDE);
@@ -1344,12 +1342,12 @@ compute_depth (int group)
   min_copper_group = MIN (bottom_group, top_group);
   max_copper_group = MAX (bottom_group, top_group);
   num_copper_groups = max_copper_group - min_copper_group;// + 1;
-  middle_copper_group = min_copper_group + num_copper_groups / 2;
+//  middle_copper_group = min_copper_group + num_copper_groups / 2;
 
   if (group >= 0 && group < max_group) {
     if (group >= min_copper_group && group <= max_copper_group) {
       /* XXX: IS THIS INCORRECT FOR REVERSED GROUP ORDERINGS? */
-      depth = -(group - middle_copper_group) * HACK_BOARD_THICKNESS / num_copper_groups;
+      depth = -(group - min_copper_group) * (HACK_BOARD_THICKNESS + HACK_COPPER_THICKNESS) / num_copper_groups;
     } else {
       depth = 0;
     }
@@ -1535,7 +1533,11 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
 
   for (group = min_phys_group; group <= max_phys_group; group++) {
 
-    Coord depth = compute_depth (group);
+#ifdef REVERSED_PCB_CONTOURS
+    Coord depth = compute_depth (group) - HACK_BOARD_THICKNESS;
+#else
+    Coord depth = compute_depth (group) + HACK_BOARD_THICKNESS / 2;
+#endif
     info.poly = NULL;
 
     fprintf (stderr, "Computing copper geometry for group %i\n", group);
@@ -1553,7 +1555,7 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
     END_LOOP;
 #endif
 
-#if 0
+#if 1
     fprintf (stderr, "Accumulating pin + via pads\n");
     r_search (PCB->Data->pin_tree, &bounds, NULL, pv_copper_callback, &info);
     r_search (PCB->Data->via_tree, &bounds, NULL, pv_copper_callback, &info);
@@ -1594,6 +1596,12 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
    *
    */
 
+#if 1
+    fprintf (stderr, "Subtracting pin + via drills\n");
+    r_search (PCB->Data->pin_tree, &bounds, NULL, pv_drill_callback, &info);
+    r_search (PCB->Data->via_tree, &bounds, NULL, pv_drill_callback, &info);
+#endif
+
   if (info.poly == NULL)
     {
       fprintf (stderr, "Skipping layer group %i, info.poly was NULL\n", group);
@@ -1606,14 +1614,14 @@ object3d_from_copper_layers_within_area (POLYAREA *area)
                               depth,                         /* Bottom */
                               depth + HACK_COPPER_THICKNESS, /* Top */
 #else
-                              -depth + HACK_COPPER_THICKNESS, /* Bottom */
-                              -depth,                         /* Top */
+                              -depth,                         /* Bottom */
+                              -depth - HACK_COPPER_THICKNESS, /* Top */
 #endif
                               copper_appearance,
                               NULL));
 
 
-    break;
+//    break;
   }
 
 
diff --git a/src/hid/step/step.c b/src/hid/step/step.c
index a4fe3c5..2bb9bfd 100644
--- a/src/hid/step/step.c
+++ b/src/hid/step/step.c
@@ -152,7 +152,7 @@ step_do_export (HID_Attr_Val * options)
   if (filename == NULL)
     filename = "pcb-out.step";
 
-#if 0
+#if 1
   board_outline_list = object3d_from_board_outline ();
 #else
   board_outline_list = NULL;
@@ -202,11 +202,13 @@ step_do_export (HID_Attr_Val * options)
         poly_DelContour (&curc);
       }
 
-//    mask_objects = object3d_from_soldermask_within_area (piece, TOP_SIDE);
-//    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+#if 1
+    mask_objects = object3d_from_soldermask_within_area (piece, TOP_SIDE);
+    board_outline_list = g_list_concat (board_outline_list, mask_objects);
 
-//    mask_objects = object3d_from_soldermask_within_area (piece, BOTTOM_SIDE);
-//    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+    mask_objects = object3d_from_soldermask_within_area (piece, BOTTOM_SIDE);
+    board_outline_list = g_list_concat (board_outline_list, mask_objects);
+#endif
 
     copper_layer_objects = object3d_from_copper_layers_within_area (piece);
     board_outline_list = g_list_concat (board_outline_list, copper_layer_objects);
